src/txrequest.h

```diff
@@ -0,0 +1,195 @@
+ *   are honest, where we assume that the attacker can disconnect and reconnect those connections, the distribution
+ *   becomes k ~ P + NB(p=1-NP/NPh,r=1) (where NB stands for Negative Binomial distribution), which has mean
+ *   P-1+NP/NPh.
+ *
+ * Complexity:
+ * - Memory usage is proportional to the total number of tracked announcements (Size()) plus the number of
+ *   peers with a nonzero number of tracked announcements.
+ * - CPU usage is generally logarithmic in the total number of tracked announcements, plus the number of
+ *   announcements affected by an operation (amortized O(1) per announcement).
+ */
+class TxRequestTracker {
+    // Avoid littering this header file with implementation details.
+    class Impl;
+    const std::unique_ptr<Impl> m_impl;
+
+public:
+    //! Construct a TxRequestTracker.
+    TxRequestTracker(bool deterministic = false);
+    ~TxRequestTracker();
+
+    // Conceptually, the data structure consists of a collection of entries, one for each peer/txhash combination
+    // (an "announcement"):
+    //
+    // - CANDIDATE entries represent transactions that were announced by a peer, and that become available for
+    //   download after their reqtime has passed.
+    //
+    // - REQUESTED entries represent transactions that have been requested, and which we're awaiting a response for
+    //   from that peer. Their expiry value determines when the request times out.
+    //
+    // - COMPLETED entries represent transactions that have been requested from a peer, and a NOTFOUND or a
+    //   transaction was received in response (valid or not), or they timed out. They're only kept around to
+    //   prevent requesting them again. If only COMPLETED entries for a given txhash remain (so no CANDIDATE or
+    //   REQUESTED ones), all of them are deleted (this is an invariant, and maintained by all operations below).
+    //
+    // The operations below manipulate the data structure.
+
+    /** Deletes all entries for a given peer.
+     *
+     * It should be called when a peer goes offline.
+     */
+    void DisconnectedPeer(uint64_t peer);
+
+    /** Deletes all entries for a given txhash (both txid and wtxid ones).
+     *
+     * This should be called when a transaction is successfully added to the mempool, seen in a block, or for
+     * whatever reason we no longer care about it.
+     */
+    void ForgetTxHash(const uint256& txhash);
+
+    /** Adds a new CANDIDATE entry.
+     *
+     * Does nothing if one already exists for that (txhash, peer) combination (whether it's CANDIDATE, REQUESTED, or
+     * COMPLETED). Note that this means a second INV with the same txhash from the same peer will be ignored, even
+     * if one is a txid and the other is wtxid (but that shouldn't happen, as BIP339 requires that all announced
+     * inventory is exclusively using MSG_WTX). The new entry is given the specified preferred and reqtime values,
```

Comment:
How does entry uniqueness is enforced w.r.t to wtxid/txid ? For segwit txn, the ByPeer can't dissociate between a txid and wtxid announcement, it's different hash ?

Reply:
I think that's explained exactly in this paragraph? "if one already exists for that (txhash, peer) combination". Uniqueness is on that combination.

If the txhash is the same, it's the same. If they're not, they're not.

Reply:
I think I hurted on "Note that this means a second INV with the same txhash from the same peer will be ignored, even if one is a txid and the other is wtxid". I interpreted it "As if first is txid, does nothing even if a second announcement is wtxid". That's my English here, nevermind.

Reply:
Just so we're on the same page. The scenario is:
* Peer P announces txid H
* Peer P announces wtxid H

In this case, the second announcement is ignored, because one already exists for peer/txhash combination (P, H). This is harmless for two reasons:
* The txid and wtxid being identical implies it's a non-segwit transaction, so it doesn't matter how we fetch.
* BIP339 prescribes that all announcements have to be wtxid when enabled (though I now realize that orphan fetching could actually mean this can actually still occur; I'll improve the comment).

Reply:
I've rewritten the comment.

Reply:
Yes clearer.